src/gdcmodels/__init__.py

```diff
@@ -1,62 +1,239 @@
-    descriptions = load_yaml(filename)
-    result = {}
-    # we can select what properties to add in the future
-    if "_meta" in descriptions:
-        result["_meta"] = descriptions.get("_meta")
-    return result
 
+class _MappingDetail(NamedTuple):
+    """The various details needed to load a given mapping."""
 
-def get_es_models(es_model_dir=None):
+    index_name: str
+    doc_type: str
+    mapping: abc.Traversable
+    settings: abc.Traversable
+    descriptions: abc.Traversable
+    vestigial: abc.Traversable
+
+
+def _extract_details(models: abc.Traversable) -> Iterator[_MappingDetail]:
+    """Extracts the mapping details from the models resource.
+
+    This works with the following model/index file structures within the given models
+    directory:
+
+    1.
+        index/
+            mapping.yaml
+            settings.yaml?
+            vestigial.yaml
+            descriptions.yaml?
+    2.
+        index/
+            doc_type/
+                mapping.yaml
+                vestigial.yaml
+            settings.yaml?
+            descriptions.yaml?
+
+    Args:
+        models: The gdcmodes es-models resource from which to extract the mapping
+            details.
+
+    Yields:
+        The details associated with each model/mapping found within the models resource.
+    """
+
+    def _extract_detail(
+        index_name: str,
+        settings: abc.Traversable,
+        descriptions: abc.Traversable,
+        doc_type: abc.Traversable,
+    ) -> _MappingDetail:
+        mapping = doc_type / "mapping.yaml"
+        vestigial = doc_type / "vestigial.yaml"
+
+        return _MappingDetail(
+            index_name, doc_type.name, mapping, settings, descriptions, vestigial
+        )
+
+    for index in models.iterdir():
+        index_name = index.name
+        settings = index / "settings.yaml"
+        descriptions = index / "descriptions.yaml"
+        extract_detail = functools.partial(_extract_detail, index_name, settings, descriptions)
+
+        # If mapping.yaml is a file then we have struct 1 and the index file can be
+        # considered the doc_type
+        if index.joinpath("mapping.yaml").is_file():
+            yield extract_detail(index)
+
+        # Otherwise, we have struct 2 and each sub dir is a doc_type
+        else:
+            doc_types = (p for p in index.iterdir() if p.is_dir())
+
+            yield from map(extract_detail, doc_types)
+
+
+def _extract_es_mapping(detail: _MappingDetail, vestigial_included: bool) -> ESMapping:
+    """Extract the elasticsearch mapping described in the given detail.
+
+    Args:
+        detail: The detail with the relavant paths from which to load the mapping.
+        vestigial_included: A flag which determins if the vestial properties should be
+            included when loading the mapping.
+
+    Returns:
+        The ESMapping loaded from the paths within the given detail.
+    """
+    mapping = yaml.safe_load(detail.mapping.read_bytes())
```

Comment:
Not super important but should we use this opportunity to use the C loaders? They are a bit faster.
```suggestion
    mapping = yaml.load(detail.mapping.read_bytes(), Loader=yaml.CSafeLoader)
```